### PR TITLE
Enrich StreamReader and SinkWriter

### PR DESCRIPTION
**Pull request by nicflower**
Implementing feature suggested in issue #5745

Basically what was done is:
- implement Sink for StreamReader<Sink>
- implement Stream for SinkWriterStream>
- implement AsyncRead for SinkWriter<AsyncRead>
- implement Stream for CopyToBytes<Stream>

The implementation is pretty straightforward since the actual work is all demanded to the underlying object, so I thought that introducing new tests would not be really helpful.

### CODE REVIEW

File: tokio-util/src/io/copy_to_bytes.rs [hunk 1]
```diff
@@ -66,3 +67,14 @@ where
...
+impl<S, T> Stream for CopyToBytes<S>
+where
+    S: Stream<Item = T> + std::marker::Unpin,
+{
+    type Item = T;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let ref_mut = self.get_mut().get_mut();
+        Pin::new(ref_mut).poll_next(cx)
+    }
+}
```
Comment (Darksonn) [inline tokio-util/src/io/copy_to_bytes.rs:80]: We don't want `Unpin` bounds if they're not necessary. Also, the `T` parameter can be dropped.
```suggestion
impl<S: Stream> Stream for CopyToBytes<S> {
    type Item = S::Item;
    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        self.project().inner.poll_next(cx)
    }
}
```

Comment (nicflower) [reply tokio-util/src/io/copy_to_bytes.rs]: fixed

Comment (Darksonn) [reply tokio-util/src/io/copy_to_bytes.rs]: Please also fix it for `SinkWriter` and `StreamReader`.

Comment (nicflower) [reply tokio-util/src/io/copy_to_bytes.rs]: done

**Review by Darksonn — APPROVED** [review]
Thanks, looks good to me!

**Pull request merged**